### DISCUSSION TITLE
Failed to connect to ESP32: Timed out waiting for packet header

### DISCUSSION THREAD

**Discussion by daveclifford** [DCC-EX/EX-T3-WiFi#47] [General]
[(solved-failed-to-connect-to-esp32-timed-out-waiting-for-packet-header](https://randomnerdtutorials.com/solved-failed-to-connect-to-esp32-timed-out-waiting-for-packet-header/)

I am willing to try this solution, so that I don't have to push the boot button each time (I don't know if the FireBeetle2 has a boot button????). I want to be able to consistently upload after the case is buttoned up.

I will report back if it fixes my upload issue using Linux. I think that the Windows drivers (that are supported) had more testing/fixes.

**Comment by daveclifford** [comment]
I ate lunch and gave this solution some thought.

The capacitor is storing an electrical charge to get over "burps" in power from the computer source.

I think my issue is due to USB/serial I/F handshaking and signals not sequencing properly by time.

It's worth a try

**Reply by alex-code** [reply]
It's got to be a software issue as it works fine on Windows. I'm using WSL with usbip.

Not sure if this'll be related to your issue https://github.com/espressif/esptool/issues/653

**Reply by daveclifford** [reply]
You are going to love what I discovered watching ESP32 YouTube videos. 

Subject: **Timing Errors And Failures  When Trying to Upload The Build**

 "To ensure the power supply to the ESP32 chip during power-up, it is advised to add an RC delay circuit at the EN pin. The recommended setting for the RC delay circuit is usually R = 10 kΩ and C = 1 µF. "

See p.13 Section 2.2.2 Reset of [https://www.espressif.com/sites/default/files/documentation/esp32_hardware_design_guidelines_en.pdf]

Some solutions show to put a small rc circuit{ R = 10 kΩ and C = 1 µF} to reduce the timing.
[https://www.reddit.com/r/esp32/comments/ou6r8d/delay_circuit_for_en_pin_calculation_of_r_and_c/]

One solution is putting a 10 UF (microfarad) capacitor between the EN Pin and Ground. The resistor  for the rc circuit is already on the board.

I watched 6 videos of various solutions, so the issue is not as rare as I thought. 

I have several Firebeetle2 ESP32 WROOM micros, so I can **sacrifice one for science**.

I'd rather add code to Platformio to determine:  1) I'm using Linux Platformio, and 2) do PRE Upload code to reset the serial lines, etc........ 

Dave

**Reply by alex-code** [reply]
Were the users in the videos using Linux?
I've never had an issue using Windows 11.

**Reply by daveclifford** [reply]
Alex,

Some were using Apple, others Windows. My guess is the source USB interface is not dumping enough power to the FireBeetle (Just an educated guess)

What surpriised me were the large (>10) number of videos on the subject.  

What 'puter are you using to run Windows?

Mine is a Dell E6420 - not the fastest, but adequate for my needs.

I believe that two issues are:  getting the CH340 driver installed in the O/S to interface to/from  the FireBeetle2 and computer and getting the permissions correct for the O/S to interface are the greatest key to success.

I bought a new USB-A to USB-C cable to ensure I had data as well as power.

I'm just eliminating all the issues

Dave

**Reply by alex-code** [reply]
Guess I've been lucky so I've not had to research on it, makes it tricky for me to help as I won't be able to reproduce to test.

I've got an old Dell XPS13.
Only ever had an odd issue where after uploading the screen didn't reset (a few times out of the 100's of uploads) so I had to just do an off-on with the power switch.

Does your laptop have other USB ports you can try?

**Reply by alex-code** [reply]
Checked my laptop and it's a USB3 port.

**Comment by daveclifford** [comment]
Well Alex,

It gets different all the time!!!!!

I am breaking out my HP Pavilion Laptop  (known on my LAN as HP15)  Model 15-cc563st with 1 USB 3.1 Type-C™ Gen 1 (Data up to 5 Gb/s) and 2 USB 3.1 Gen 1 (Data Transfer Only). It is around 5 years old.

I have to update the USB CH340 driver and permissions, but, maybe - just maybe,  the USB port might work with the FireBeetle2.

The keyboard has square keys [and the key pressure is different] - but the keys are lit. The delete key is in 2  places (not what my brain likes).

I'l be working on this tomorrow and will report back - either way!

The Linux implementation of VSC and PlatformIO use a python-based  ESPTool  app (3K+) to control the USB serial control.

[Changing the subject] You short the EN/GND pins to turn off (or is it on?) the throttle, so a 10UF capacitor across them may work (my last resort).

I'd much rather let the 'puter USB power fix the issue.

Dave

**Reply by alex-code** [reply]
Pulling EN to GND turns off the regulator (but still allows charging).

**Comment by daveclifford** [comment]
Alex,

I had this crazy thought re: the upload failing for me. I am trying to upload the build w/o having a battery attached. I only have the components on the frame w/o wiring and no battery attached.

**Maybe the battery provides power and filtering to the ESP32, such that the microprocessor can change state to upload.**

I'll stick the battery on, let it charge wnd then try the upload.

Worth a try before adding more hardware ( rc circuit).


-------------------------------------------------------

Flashing
 [¶](https://docs.espressif.com/projects/arduino-esp32/en/latest/troubleshooting.html#flashing)

Why is my board not flashing/uploading when I try to upload my sketch?

[¶](https://docs.espressif.com/projects/arduino-esp32/en/latest/troubleshooting.html#why-is-my-board-not-flashing-uploading-when-i-try-to-upload-my-sketch)

To be able to upload the sketch via serial interface, the ESP32 must be in the download mode. The download mode allows you to upload the sketch over the serial port and to get into it, you need to keep the GPIO0 in LOW while a resetting (EN pin) cycle. If you are trying to upload a new sketch and your board is not responding, there are some possible reasons.

Possible fatal error message from the Arduino IDE:

A fatal error occurred: Failed to connect to ESP32: Timed out waiting for packet header

Solution[¶](https://docs.espressif.com/projects/arduino-esp32/en/latest/troubleshooting.html#id1)

Here are some steps that you can try to:

Check your USB cable and try a new one.

Change the USB port.

**Check your power supply.**

In some instances, you must keep GPIO0 LOW during the uploading process via the serial interface.

Hold down the “BOOT” button in your ESP32 board while uploading/flashing.

In some development boards, you can try adding the reset delay circuit, as described in the Power-on Sequence section on the [ESP32 Hardware Design Guidelines](https://www.espressif.com/sites/default/files/documentation/esp32_hardware_design_guidelines_en.pdf) in order to get into the download mode automatically.

**Reply by alex-code** [reply]
Do you have access to a computer with a USB3 port?
I can successfully flash with or without a battery on different computers, the only common theme is they all had USB3.

USB3 can give more mA, 900mA vs 500mA for USB2.

**Comment by daveclifford** [comment]
The HP15 has a USB3 port.

The issue is that I have to add the CH340 driver, USB port and permissions
to the :Linux O/S on the HP15.

Stay tuned

Dave

**Reply by alex-code** [reply]
I'm not that familiar with Linux.
Googling for CH340 I found an article by SparkFun.
https://learn.sparkfun.com/tutorials/how-to-install-ch340-drivers/linux

**Comment by daveclifford** [comment]
How about this?

**Insufficient Power**

[(https://docs.espressif.com/projects/esptool/en/latest/esp32/troubleshooting.html#insufficient-power)]

The 3.3V power supply for the ESP chip has to supply large amounts of current (up to 70mA continuous, **200-300mA peak,** might be slightly higher). You also **need sufficient capacitance on the power circuit to meet large spikes of power demand.**

Dave

**Reply by daveclifford** [reply]
I'm breaking down and buying a - DROK USB Tester, USB-A USB-C 2.0 Multimeter Dual USB Ports due in tomorrow. 

I'll let you know the measurements on both the  E6420 Dell laptop - **USB-2** (~10 years) and the HP15 (~7 years) **USB-C** .

Dave

**Reply by daveclifford** [reply]
Today is SuperBowl Sunday in the USA.

My Drok  USB Tester came last PM

With the FireBeetle2 plugged into the E6420, the voltage draw is 5.04V; the current draw fluctuates between 7MA and 17MA.
(It went to 33MA  just a moment ago!!)

Next, I am going to use a contact cleaner to clean  the USB port,

The upload has worked two times. without modification, so, either the timing was fortuitous, or the USB connector made the right  contact (the USB sockets are hardwired to the laptop motherboard, so a replacement port could take alot of effort to fix).

Dave

**Comment by daveclifford** [comment]
No battery, no special caps..  
I do NOT have the SEN0235 sor and SEN0405 wired - The screen flashs on/off at around 500 ms intervals. 
I will wire them in tomorrow.

Some progress - the log shows an upload occured.  (I only  prayed...................)

 --------------------- upload log below -------------------------------------------

Looking for upload port...
Using manually specified: /dev/ttyUSB0
**Uploading .pio/build/firebeetle2/firmware.bin**
**esptool.py v4.4
Serial port /dev/ttyUSB0**
Connecting.............................
Chip is ESP32-D0WD-V3 (revision v3.0)
Features: WiFi, BT, Dual Core, 240MHz, VRef calibration in efuse, Coding Scheme None
Crystal is 40MHz
MAC: 08:3a:f2:39:0e:24
Uploading stub...
Running stub...
Stub running...
Changing baud rate to 460800
Changed.
Configuring flash size...
Flash will be erased from 0x00001000 to 0x00005fff...
Flash will be erased from 0x00008000 to 0x00008fff...
Flash will be erased from 0x0000e000 to 0x0000ffff...
Flash will be erased from 0x00010000 to 0x00155fff...
Compressed 17408 bytes to 12112...
**Writing at 0x00001000... (100 %)
Wrote 17408 bytes (12112 compressed) at 0x00001000 in 0.5 seconds (effective 263.8 kbit/s)...
Hash of data verified.**
Compressed 3072 bytes to 134...
Writing at 0x00008000... (100 %)
Wrote 3072 bytes (134 compressed) at 0x00008000 in 0.1 seconds (effective 460.9 kbit/s)...
Hash of data verified.
Compressed 8192 bytes to 47...
Writing at 0x0000e000... (100 %)
Wrote 8192 bytes (47 compressed) at 0x0000e000 in 0.1 seconds (effective 640.5 kbit/s)...
Hash of data verified.
Compressed 1332768 bytes to 797668...
Writing at 0x00010000... (2 %)
Writing at 0x0001a728... (4 %)
Writing at 0x000252b1... (6 %)
Writing at 0x0002deda... (8 %)
Writing at 0x0003afe1... (10 %)
Writing at 0x00050297... (12 %)
Writing at 0x00056062... (14 %)
Writing at 0x0005c647... (16 %)
Writing at 0x000629fb... (18 %)
Writing at 0x00069490... (20 %)
Writing at 0x0006ea01... (22 %)
Writing at 0x00074353... (24 %)
Writing at 0x0007a569... (26 %)
Writing at 0x0008015f... (28 %)
Writing at 0x00086170... (30 %)
Writing at 0x0008bdfa... (32 %)
Writing at 0x00091701... (34 %)
Writing at 0x0009749e... (36 %)
Writing at 0x0009c8e8... (38 %)
Writing at 0x000a1a37... (40 %)
Writing at 0x000a6e2f... (42 %)
Writing at 0x000ac23f... (44 %)
Writing at 0x000b1414... (46 %)
Writing at 0x000b6393... (48 %)
Writing at 0x000bb82c... (51 %)
Writing at 0x000c0c49... (53 %)
Writing at 0x000c5eb1... (55 %)
Writing at 0x000cbbc1... (57 %)
Writing at 0x000d1342... (59 %)
Writing at 0x000d6486... (61 %)
Writing at 0x000db8a2... (63 %)
Writing at 0x000e0c3c... (65 %)
Writing at 0x000e6133... (67 %)
Writing at 0x000eb917... (69 %)
Writing at 0x000f1897... (71 %)
Writing at 0x000f7330... (73 %)
Writing at 0x000fd89f... (75 %)
Writing at 0x001051e6... (77 %)
Writing at 0x0010ec83... (79 %)
Writing at 0x00115868... (81 %)
Writing at 0x0011bce7... (83 %)
Writing at 0x00123bd3... (85 %)
Writing at 0x0012c88b... (87 %)
Writing at 0x00134c63... (89 %)
Writing at 0x0013a746... (91 %)
Writing at 0x00141609... (93 %)
Writing at 0x00146dae... (95 %)
Writing at 0x0014c0d4... (97 %)
Writing at 0x001516bc... (100 %)
**Wrote 1332768 bytes (797668 compressed) at 0x00010000 in 19.7 seconds (effective 540.6 kbit/s)...
Hash of data verified.**

Leaving...
Hard resetting via RTS pin...

**Reply by alex-code** [reply]
You need to upload/flash the file system too.

**Reply by daveclifford** [reply]
Alex,

Can you clarify the steps we need to take to get the code from the PlatformIO tool to the FireBeetle 2?

Assuming there is a new update to your code:

In the folloiwing order:

Under Project Tasks (Firebeetle2/General)

1. Build
2. Upload (or Upload and Monitor)

Under Project Tasks (Firebeetle2/Platform)

1. Build Filesystem Image
2. Upload Filesystem Image

Questions:

1. How often do we need to do this -  after every code update? 
2. Does it matter the sequence (General Tasks, then Platform Tasks, or Platform Tasks, then General Tasks)  ???

Burning questions your fans want to know!!

(I confess I never read the Code.md - I should have!!!!!!!; I was more excited by the Assembly.md words and colored pictures!!!!!)

**Reply by alex-code** [reply]
Order doesn't matter too much but the code won't work without the file system being uploaded.

I'm working towards versioned releases so you'd only need to update with a new release.
Once the merged builds are working you'd only need one upload instead of separate ones too.

**Reply by daveclifford** [reply]
Ooohhhhhh Goodie!

I'm going to put in a capacitor to  see if it helps the upload issue
Dave

**Comment by daveclifford** [comment]
Alex,

I did NOT put the capacitor between the FireBeetle2 EN and Ground points.

Instead, I picked up the HP15 notebook, d/led the newest EX-T3-WIFI build, updated VSCode and PlatformIO and then checked out my development environment - the CH340 Serial/USB driver, permissions,etc.

After reading (your response) that the Build/Upload sequence didn't matter, (with baited breath and fingers and toes crossed),  I was able to Build and Upload both the General and Platform code.

The screen lit up onto  the WiFi screen (not flashing) and I was able to add the WiFi Name and P/W combo.

I have to wire the Rotary Encoder and Accelerometer to the FireBeetle2, so I couldn't get to any other screens.

GREAT PROGRESS!!!

Thank you (so much) for your help and responses !!!!

Lessons Learned:

1.  The E6420 should have worked - I will spend some time cleaning the USB-A  port, but, since the HP15 (Linux) notebook works fine, I am no longer impatient.
2. I may have wasted time troubleshooting the E6420 uploading, but I now know Linux can be used as well as MSWindows for development.
3. I should have read the code.md file (It should be called **Upload Code.md**).

I will add some pics in a bit.  I am building 2 Wi-FI Throttles - this first one  is a test run - of sorts!

I am also building a DCC-EX Arduino/Motor Shield/WiFi system.

Dave
